packages/flutter/lib/src/widgets/magnifier.dart

```diff
@@ -0,0 +1,545 @@
+              child: SizedBox.fromSize(
+                size: size,
+                child: child,
+              ),
+            ),
+        ),
+
+      // Because `BackdropFilter` will filter any widgets before it, we should
+      // apply the style after (i.e. in a younger sibling) to avoid the magnifier
+      // from seeing its own styling.
+        Opacity(
+          opacity: decoration.opacity,
+          child: _MagnifierStyle(
+              decoration,
+              size: size,
+          ),
+        )
+      ],
+    );
+  }
+}
+
+class _MagnifierStyle extends StatelessWidget {
+  const _MagnifierStyle(this.decoration, {required this.size});
+
+  final MagnifierDecoration decoration;
+  final Size size;
+
+  @override
+  Widget build(BuildContext context) {
+    double largestShadow = 0;
+    for (final BoxShadow shadow in decoration.shadows ?? <BoxShadow>[]) {
+      largestShadow = math.max(
+          largestShadow,
+          (shadow.blurRadius + shadow.spreadRadius) +
+              math.max(shadow.offset.dy.abs(), shadow.offset.dx.abs()));
+    }
+
+    return ClipPath(
+      clipBehavior: Clip.hardEdge,
+      clipper: _DonutClip(
+        shape: decoration.shape,
+        spreadRadius: largestShadow,
+      ),
+      child: DecoratedBox(
+        decoration: decoration,
+        child: SizedBox.fromSize(
+          size: size,
+        ),
+      ),
+    );
+  }
+}
+
+/// A `clipPath` that looks like a donut if you were to fill its area.
+///
+/// This is necessary because the shadow must be added after the magnifier is drawn,
+/// so that the shadow does not end up in the magnifier. Without this clip, the magnifier would be
+/// entirely covered by the shadow.
+///
+/// The negative space of the donut is clipped out (the donut hole, outside the donut).
+/// The donut hole is cut out exactly like the shape of the magnifier.
+class _DonutClip extends CustomClipper<Path> {
+  _DonutClip({required this.shape, required this.spreadRadius});
+
+  final double spreadRadius;
+  final ShapeBorder shape;
+
+  @override
+  Path getClip(Size size) {
+    final Path path = Path();
+    final Rect rect = Offset.zero & size;
+
+    path.fillType = PathFillType.evenOdd;
+    path.addPath(shape.getOuterPath(rect.inflate(spreadRadius)), Offset.zero);
+    path.addPath(shape.getInnerPath(rect), Offset.zero);
+    return path;
+  }
+
+  @override
+  bool shouldReclip(_DonutClip oldClipper) => oldClipper.shape != shape;
+}
+
+class _Magnifier extends SingleChildRenderObjectWidget {
+  /// Construct a [_Magnifier].
+  const _Magnifier({
+      super.child,
+      required this.shape,
+      this.magnificationScale = 1,
+      this.focalPoint = Offset.zero,
+  });
+
+  /// [focalPoint] is the area the center of the
+  /// [_Magnifier] points to, relative to the center of the magnifier.
+  /// If left as [Offset.zero], the magnifier will magnify whatever is directly
+  /// below it.
+  final Offset focalPoint;
```

Comment:
Just want to make sure I'm understanding this correctly: this determines where the transformed `BackdropLayer`  appears, not which part of the content below the magnifier is shown? IOW if I keep increasing the `Offset` value but at the same time keep everything else in the app the same, the image that appears in the magnifier would remain the same, but the location of the magnifier would move farther away from the origin?

Reply:
Ah looking at `addToScene` the clipping isn't transformed so I guess this points to the content you want to show in the magnifier? And positive (dx, dy) means the magnifier will show things that are actually on the top left corner of the app?

Reply:
Close: it's the offset from the center of the magnifier. For instance, if the magnifier is at (100, 100), and focalPoint is 20, 20, you'd see (120, 120) in the magnifier

Reply:
Elaborated and also from an offline conversation: the logic was actually flipped. Passing in 20, 20 would get you 80, 80 in this scenario, since we're transforming the background. I made it do what it sounds like (just multiplied the focal point by -1 at the end, and fixed all the calculations to pass in the negated value)